File: c/sieve-of-sundaram.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int sieve_of_sundaram(int **primes, int limit) {
	int new_limit = (limit - 1) / 2;
	char *array = malloc(new_limit);
	for (int i = 0; i < new_limit; ++i) {
		array[i] = 0;
	}
	// sieve out the matrix
	int i = 1;
	int j = i;
	while (2 * i * j + i + j <= new_limit) {
		while(2 * i * j + i + j <= new_limit) {
			array[2 * i * j + i + j - 1] = 1;
			++j;
		}
		++i;
		j = i;
	}
	// count the primes
	int prime_count = 0;
	for (int i = 0; i < new_limit; ++i) {
		if (array[i] == 0) {
			++prime_count;
		}
	}
	// fill in primes[] with the primes
	*primes = malloc(++prime_count * sizeof(int));
	i = 0;
	*primes[i++] = 2;
	for (int j = 0; j < new_limit; ++j) {
		if (array[j] == 0) {
			(*primes)[i++] = j * 2 + 3;
		}
	}
	return prime_count;
}

int main(int argc, char const *argv[]) {
	int limit = 100;
	if (argc == 2)
		limit = atoi(argv[1]);

	clock_t t;
	t = clock();

	int *primes;
	int prime_count = sieve_of_sundaram(&primes, limit);

	t = clock() - t;
	double time_taken = ((double)t)/CLOCKS_PER_SEC;
	printf("The program took %f seconds to execute.\n", time_taken);
	printf("It computed %d primes, %f primes per second\n", prime_count, prime_count / time_taken);
	return 0;
}

File: python/sieve_of_sundaram.py
import time
import math
import sys
time_beginning = time.time()
def sieve_of_sundaram(limit):
	new_limit = (limit - 1) // 2
	boolean_list = [True] * new_limit;
	for i in range(1, int((math.sqrt(2 * limit + 1) - 1) / 2) + 1):
		start = 2 * i + 2 * i * i
		for j in range(start, new_limit + 1, 2 * i + 1):
			boolean_list[j - 1] = False
			j += 1
	primes = [2]
	for i in range(0, new_limit):
		if boolean_list[i]:
			primes.append(i * 2 + 3)
	return primes

primes = sieve_of_sundaram(int(sys.argv[1]))
print(len(primes))
time_taken = time.time() - time_beginning
print(str(time_taken))


# in the first for loop at line 10, the upper boundary for the range is derived
# by solving for 2 * x ** 2 + 2 * x < limit - > x is limited by
# (sqrt(2 * limit + 1) - 1) / 2
# the second loop is a direct transformation